xen/arch/x86/boot/head.S

```diff
@@ -368,12 +416,13 @@ __start:
         /* Restore the clobbered field. */
         mov     %edx, (%ebx)
 
+.Lset_stack:
         /* Set up stack. */
         lea     STACK_SIZE - CPUINFO_sizeof + sym_esi(cpu0_stack), %esp
 
         /* Initialize GDTR and basic data segments. */
-        add     %esi, sym_esi(gdt_boot_base)
-        lgdt    sym_esi(gdt_boot_descr)
+        add     %esi, %cs:sym_esi(gdt_boot_base)
+        lgdt    %cs:sym_esi(gdt_boot_descr)
```

Comment:
Code segments are read-only, `add` causes segfault, so the second line doesn't need to change either.

Reply:
I haven't thought about it. Then again, according to TXT Software Design Guide, on entry to MLE (Xen) DS, ES and SS are undefined. This makes it impossible to load new GDT without having the descriptor that would already have a proper base address in it...

Data segments were valid for all CPUs we've tested so far, so maybe the best option is to assume that this is an error in the documentation and ACM actually sets these segments. It has to somehow access the data anyway.

Reply:
Undid that change and explained assumption in the comment: https://github.com/TrenchBoot/xen/commit/6d994f475c28b7b84fc89cf091ccc3599283ff4b#diff-e4b4491b6c4e847098c559822abc0da4b62e431b9339a971a79026a0c3bac48aR361-R365